src/autopas/utils/SimilarityFunctions.h

```diff
@@ -28,86 +46,68 @@ template <class Container>
+  std::vector<size_t> particlesPerBin(numberOfBins, 0);
 
-  // add particles accordingly to their cell to get the amount of particles in each cell
+  // add particles accordingly to their bin to get the amount of particles in each bin
   for (auto particleItr = container.begin(autopas::IteratorBehavior::owned); particleItr.isValid(); ++particleItr) {
     const auto &particleLocation = particleItr->getR();
-    std::array<size_t, 3> index = {};
-    for (size_t i = 0; i < particleLocation.size(); i++) {
-      index[i] = particleLocation[i] / cellLength;
-    }
-    const size_t cellIndex = autopas::utils::ThreeDimensionalMapping::threeToOneD(index, cellsPerDimension);
-    particlesPerCell[cellIndex] += 1;
-    // calculate the size of the current cell
-    allVolumes[cellIndex] = 1;
-    for (size_t i = 0; i < cellsPerDimension.size(); ++i) {
-      // the last cell layer has a special size
-      if (index[i] == cellsPerDimension[i] - 1) {
-        allVolumes[cellIndex] *= outerCellSizePerDimension[i];
-      } else {
-        allVolumes[cellIndex] *= cellLength;
-      }
-    }
+
+    const auto binIndex3d = autopas::utils::ArrayMath::floorToInt((particleLocation - startCorner) / binDimensions);
+    const auto binIndex1d = autopas::utils::ThreeDimensionalMapping::threeToOneD(
+        binIndex3d, {binsPerDimension, binsPerDimension, binsPerDimension});
+
+    particlesPerBin[binIndex1d] += 1;
   }
 
-  // calculate density for each cell
+  // calculate density for each bin and track max density
   double maxDensity{0.};
-  std::vector<double> densityPerCell(numberOfCells, 0.0);
-  for (size_t i = 0; i < particlesPerCell.size(); i++) {
-    densityPerCell[i] =
-        (allVolumes[i] == 0) ? 0 : (particlesPerCell[i] / allVolumes[i]);  // make sure there is no division of zero
-    if (densityPerCell[i] > maxDensity) {
-      maxDensity = densityPerCell[i];
+  std::vector<double> densityPerBin;
+  densityPerBin.reserve(numberOfBins);
+  for (size_t i = 0; i < numberOfBins; i++) {
+    densityPerBin[i] = (double)particlesPerBin[i] / binVolume;
+    if (densityPerBin[i] > maxDensity) {
+      maxDensity = densityPerBin[i];
     }
   }
 
-  if (maxDensity < 0.0)
+  if (maxDensity < 0.0) {
     throw std::runtime_error("maxDensity can never be smaller than 0.0, but is:" + std::to_string(maxDensity));
-
+  }
   // get mean and reserve variable for densityVariance
-  const double densityMean = numberOfParticles / volume;
+  const double densityMean = numberOfParticles / domainVolume;
   double densityVariance = 0.0;
 
   // calculate densityVariance
-  for (size_t r = 0; r < densityPerCell.size(); ++r) {
-    double distance = densityPerCell[r] - densityMean;
-    densityVariance += (distance * distance / densityPerCell.size());
+  for (size_t i = 0; i < numberOfBins; ++i) {
+    const double densityDifference = densityPerBin[i] - densityMean;
+    densityVariance += (densityDifference * densityDifference / numberOfBins);
   }
```

Comment:
I'm not a fan of this variable + for loop pattern. Therefore, I'd replace this with some function from `algorithm.h` (`reduce` or `transform_reduce`?). But I could live with the loop if you prefer it ;)

Reply:
I personally don't really care about the style here, but in general, I often find the `transform_reduce` constructs harder to read than the for loop which does the computations similar to how I would do it by hand.
I would therefore be interested to know, why you(@FG-TUM ) prefer std algorithms whenever possible?  
(Maybe the readability thing is just matter of habit)

Reply:
Yes, I, too, think readability is more of a matter of habit. The reason why I prefer the algorithms is that it is significantly harder to sneak in suspicious behavior. For example, in an index-based loop, you can mess with the index in the loop, start or end at any number, but this might be very easy to miss in a review because you **expect** it to go from 0 to end. A std algorithm guarantees no index mess. Therefore, I prefer to use for-loops only when such shenanigans are actually necessary.

The second point is that std algorithms allow you to initialize the result variable directly and make it const, decreasing the mental complexity of the rest of the code block.

Reply:
Done (as of 92553a1e)

Reply:
Nope you did it here: b4d74789cd1e8b259c4adc931b964198c0403dfc